src/main/java/com/maddyhome/idea/vim/ui/StatusBarModeFactory.kt

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2003-2023 The IdeaVim authors
+ *
+ * Use of this source code is governed by an MIT-style
+ * license that can be found in the LICENSE.txt file or at
+ * https://opensource.org/licenses/MIT.
+ */
+
+package com.maddyhome.idea.vim.ui
+
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.project.ProjectManager
+import com.intellij.openapi.wm.StatusBarWidget
+import com.intellij.openapi.wm.StatusBarWidgetFactory
+import com.intellij.openapi.wm.WindowManager
+import com.maddyhome.idea.vim.VimPlugin
+import com.maddyhome.idea.vim.api.VimEditor
+import com.maddyhome.idea.vim.api.injector
+import com.maddyhome.idea.vim.common.EditorFocusListener
+import com.maddyhome.idea.vim.common.ModeChangeListener
+import com.maddyhome.idea.vim.newapi.globalIjOptions
+import com.maddyhome.idea.vim.state.mode.Mode
+import com.maddyhome.idea.vim.state.mode.SelectionType
+import com.maddyhome.idea.vim.state.mode.mode
+import java.awt.Component
+
+private const val ID = "IdeaVim::Mode"
+
+public class StatusBarModeFactory : StatusBarWidgetFactory {
+  private companion object {
+    private const val NO_MODE = "" // for cases were no editors are focused
```

Comment:
Why not to use getStatusString function?

Reply:
For several reasons:
1. We use a different format for names (e.g., 'INSERT' instead of '-- INSERT --').
2. This widget is only needed for the 'showmode' option. It should be defined in the location where it's needed. There's no need to include this method in the VimStateMachine interface, and I would like to remove it along with the old widget.